### PR TITLE
Implement console endpoint on slave

### PR DESCRIPTION
**Pull request by tjlee0909**
This entails implementing the console_output class, which is a utility class for getting segments of console output from a file.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @tjlee0909 has signed the CLA. Thanks for the pull request!

File: app/web_framework/cluster_slave_application.py [hunk 1]
```diff
@@ -116,6 +126,82 @@ def get(self, build_id, subjob_id):
...
+            }
+        self.write(response)
+
+
+class _AtomHandler(_ClusterSlaveBaseAPIHandler):
+    def get(self, build_id, subjob_id, atom_id):
+        response = {
+            'comment': 'not implemented',
+            }
+        self.write(response)
```
Comment (tjlee0909) [inline app/web_framework/cluster_slave_application.py]: I wasn't quite sure what to put here for these two endpoints (_AtomsHandler and _AtomHandler) that do nothing, so I just copied what we have for these types of non-implemented endpoints.

File: app/common/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+__author__ = 'taejun'
```
Comment (josephharrington) [inline app/common/__init__.py]: :dollar: :dollar: :tophat: :dollar: :dollar:

Comment (aptxkid) [inline app/common/__init__.py]: @tjlee0909  You can change this in IntelliJ settings

Comment (tjlee0909) [inline app/common/__init__.py]: Dope

**Comment by tjlee0909** [comment]
Dangit, I just realized that in PHPUnit, the output looks like: 

PHPUnit 4.0.12 by Sebastian Bergmann.

.....

And if we read output by lines, and not by bytes, then the streaming console output will not show the progressing dots (because they are all on the same line, and we have already requested that line in a previous request).

File: app/web_framework/cluster_slave_application.py [hunk 3]
```diff
@@ -116,6 +126,82 @@ def get(self, build_id, subjob_id):
...
+            response = {'message': '\'offset_line\' must be greater than or equal to zero.'}
+            self._write_status(response, success=False, status_code=400)
+            return
+        if max_lines <= 0:
+            response = {'message': '\'max_lines\' must be greater than zero.'}
+            self._write_status(response, success=False, status_code=400)
+            return
+
+        artifact_dir = BuildArtifact.atom_artifact_directory(build_id, subjob_id, atom_id)
+        output_file = os.path.join(artifact_dir, BuildArtifact.OUTPUT_FILE)
```
Comment (josephharrington) [inline app/web_framework/cluster_slave_application.py]: How do you feel about moving most of this logic in this method into `ClusterSlave`? Originally the `ClusterMasterApplication` and `ClusterSlaveApplication` classes were intended to just be thin wrappers around `ClusterMaster` and `ClusterSlave`. (The original motivation was that we should ideally be able to swap out the web framework layer without affecting underlying app behavior.)

We've moved away from that a bit and some business logic has crept in to the web framework layer, but I think it's still a good ideal to shoot for.

Something you could also take advantage of is from `ClusterSlave` raise a `BadRequestError` with the error message and the web framework layer will automatically create the `{'message': 'whatever'}` response instead of you needing to construct it manually here.

File: app/web_framework/cluster_slave_application.py (hunk 1 above)
Comment (josephharrington) [inline app/web_framework/cluster_slave_application.py]: Could you just do a `pass` similar to `_SubjobsHandler` above? I'm not sure what the reasoning behind the original "not implemented" response was.

File: app/web_framework/cluster_slave_application.py [hunk 4]
```diff
@@ -116,6 +126,82 @@ def get(self, build_id, subjob_id):
...
+
+        # If the console output file doesn't exist where we expect it to be, return a 404.
+        if not os.path.isfile(output_file):
+            response = {'message': 'Output file doesn\'t exist for build_id: {} subjob_id: {} atom_id: {}'.format(
+                build_id, subjob_id, atom_id)}
+            self._write_status(response, success=False, status_code=404)
```
Comment (josephharrington) [inline app/web_framework/cluster_slave_application.py]: Continuing from my comment above, a 404 can be set by raising an `ItemNotFoundError`. The full list is at the top of the `ClusterBaseHandler` class.

File: test/integration/common/test_console_output.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
+from genty import genty, genty_dataset
+import os
+from tempfile import mkstemp
+from unittest import TestCase
+
+from app.common.console_output import ConsoleOutput
+
+
+@genty
+class TestConsoleOutput(TestCase):
```
Comment (josephharrington) [inline test/integration/common/test_console_output.py]: Should we create an (empty for now) `BaseIntegrationTestCase` class? I can see us definitely using that as we create more integration tests.

File: app/web_framework/cluster_slave_application.py (hunk 1 above)
Comment (tjlee0909) [inline app/web_framework/cluster_slave_application.py]: Done

File: app/web_framework/cluster_slave_application.py (hunk 4 above)
Comment (tjlee0909) [inline app/web_framework/cluster_slave_application.py]: That's cool, will use that.

File: app/web_framework/cluster_slave_application.py (hunk 3 above)
Comment (tjlee0909) [inline app/web_framework/cluster_slave_application.py]: Sounds good.

File: test/integration/common/test_console_output.py (hunk 5 above)
Comment (tjlee0909) [inline test/integration/common/test_console_output.py]: OK

File: app/common/console_output.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        :param path: sys path to the console output file
+        :type path: str
+        """
+        self.path = path
+
+    def segment(self, max_lines=50, offset_line=None):
+        """
+        Return a segment of the console output.
+
+        :param max_lines:
```
Comment (josephharrington) [inline app/common/console_output.py]: It's pretty obvious what this is supposed to be but did you mean to add a docstring for the param?

Comment (tjlee0909) [inline app/common/console_output.py]: Nah I'll delete it.

File: app/common/console_output.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                        offset_line, self.path))
+
+                total_lines += 1
+
+            # Retrieve the console_output just between offset_line and offset_line + max_lines
+            for i in range(offset_line, offset_line + max_lines):
+                line = f.readline()
+
+                # We have reached the end of the file, or a line that has not finished being written to.
+                if line == '' or not line.endswith("\n"):
```
Comment (josephharrington) [inline app/common/console_output.py]: Just curious -- is there an easy way to manually test the case where a line has not finished being written to? I tried testing a one-line file (e.g., `echo something > oneline.txt`) but Python still sticks a newline on the end.

File: test/integration/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+__author__ = 'taejun'
```
Comment (tjlee0909) [inline test/integration/__init__.py]: DOPE

File: app/common/console_output.py (hunk 7 above)
Comment (tjlee0909) [inline app/common/console_output.py]: I modified a PHPUnit test to perform a "sleep(60)" on the third test case, and did this:

phpunit test/unit/lib/Jira_Test.php > whatever.txt

and then cntrl-c'd out of it a few seconds into it.
```
taejun@mbp-002672 devtools (dev) $ cat whatever.txt 
PHPUnit 4.0.12 by Sebastian Bergmann.

Configuration read from /Users/taejun/Code/devtools/phpunit.xml

..taejun@mbp-002672 devtools (dev) $ 
```

and I see here that the terminal prompt is on the same line as the cat output.

File: app/common/console_output.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        :rtype: ConsoleOutputSegment
+        """
+        total_lines = 0
+        output_lines = 0
+        console_output = []
+
+        with open(self.path, 'r') as f:
+            # Figure out how many lines there are, so that we know which offset line to start appending
+            # console output from.
+            for line in iter(f.readline, ''):
```
Comment (josephharrington) [inline app/common/console_output.py]: I think this is equivalent to
```python
for line in f:
    ...
```

**Comment by tjlee0909** [comment]
Aw, the coveralls doesn't include my integration test and assumes that I lowered the coverage percentage.

File: app/common/console_output.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+                if line.endswith("\n"):
+                    total_lines += 1
+
+            f.seek(0)
+            offset_line = max(total_lines - max_lines, 0)
+
+            # Move file pointer to the desired offset_line
```
Comment (josephharrington) [inline app/common/console_output.py]: Just curious -- do you see any downside to calling `_parse_from_offset` at this point?

Comment (tjlee0909) [inline app/common/console_output.py]: I was considering that. The only downside is that we close and reopen the file pointer, but that's not much.

**Comment by josephharrington** [comment]
@tjlee0909 if you change the `test-unit-with-coverage` target in the Makefile to also run the integration test directory it should include it in the coverage.

File: app/slave/cluster_slave.py [hunk 11]
```diff
@@ -324,6 +327,55 @@ def kill(self):
...
+        except ValueError as e:
+            raise BadRequestError(e)
+
+        return {
+            'offset_line': segment.offset_line,
+            'num_lines': segment.num_lines,
+            'total_num_lines': segment.total_num_lines,
+            'content': segment.content,
+        }
+
```
Comment (josephharrington) [inline app/slave/cluster_slave.py]: This is pretty clean and great. :smiley_cat:

File: test/integration/common/test_console_output.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+        self.assertEquals(segment.num_lines, expected_num_lines)
+        self.assertEquals(segment.offset_line, expected_offset_line)
+        self.assertEquals(segment.total_num_lines, expected_total_num_lines)
+        self.assertEquals(segment.content, expected_content)
+
+    def test_segment_raises_value_error_if_offset_greater_than_total_length(self):
+        with self.assertRaises(ValueError):
+            console_output = ConsoleOutput(self._completed_console_output_file_path)
+            console_output.segment(max_lines=5, offset_line=155)
```
Comment (josephharrington) [inline test/integration/common/test_console_output.py]: I remember we ran into some weird issues with invalid unicode coming out of phpunit output that caused some exceptions. (We have this test around the issue: https://github.com/box/ClusterRunner/blob/master/test/unit/project_type/test_project_type.py#L78) I wonder if we should add some tests for this new functionality related to unicode console output and potentially malformed output.

I think we're a bit safer in this case because we're handling a client's API call when loading the file, so the worst that could happen is returning a 500 to the client.

File: app/common/console_output.py (hunk 7 above)
Comment (josephharrington) [inline app/common/console_output.py]: Oh derp, yeah that makes sense. I forgot that echo adds a newline by default. `echo -n something > oneline.txt` works.

**Comment by josephharrington** [comment]
Looks great! I made a few comments, but overall I am happy. :smile: :+1:

File: app/common/console_output.py (hunk 9 above)
Comment (tjlee0909) [inline app/common/console_output.py]: Done.

File: app/common/console_output.py (hunk 10 above)
Comment (tjlee0909) [inline app/common/console_output.py]: I tried doing that, and for some reason it swapped back and forth between passing my integration tests and failing it. I've decided to keep this for now.
